api/src/main/java/com/okta/sdk/resource/identity/provider/IdentityProviderBuilder.java

```diff
@@ -0,0 +1,175 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.sdk.resource.identity.provider;
+
+import com.okta.commons.lang.Classes;
+import com.okta.sdk.client.Client;
+import com.okta.sdk.resource.policy.PolicySubjectMatchType;
+
+import java.util.List;
+
+public interface IdentityProviderBuilder {
+
+    interface OIDCIdentityProviderBuilder {
+        OIDCIdentityProviderBuilder setName(String name);
+
+        OIDCIdentityProviderBuilder setIssuerMode(IdentityProvider.IssuerModeEnum issuerMode);
+
+        OIDCIdentityProviderBuilder setRequestSignatureAlgorithm(String requestSignatureAlgorithm);
+
+        OIDCIdentityProviderBuilder setRequestSignatureScope(ProtocolAlgorithmTypeSignature.ScopeEnum requestSignatureScope);
+
+        OIDCIdentityProviderBuilder setResponseSignatureAlgorithm(String responseSignatureAlgorithm);
+
+        OIDCIdentityProviderBuilder setResponseSignatureScope(ProtocolAlgorithmTypeSignature.ScopeEnum responseSignatureScope);
+
+        OIDCIdentityProviderBuilder setAcsEndpointBinding(ProtocolEndpoint.BindingEnum acsEndpointBinding);
+
+        OIDCIdentityProviderBuilder setAcsEndpointType(ProtocolEndpoint.TypeEnum acsEndpointType);
+
+        OIDCIdentityProviderBuilder setAuthorizationEndpointBinding(ProtocolEndpoint.BindingEnum authorizationEndpointBinding);
+
+        OIDCIdentityProviderBuilder setAuthorizationEndpointUrl(String authorizationEndpointUrl);
+
+        OIDCIdentityProviderBuilder setTokenEndpointBinding(ProtocolEndpoint.BindingEnum tokenEndpointBinding);
+
+        OIDCIdentityProviderBuilder setTokenEndpointUrl(String tokenEndpointUrl);
+
+        OIDCIdentityProviderBuilder setUserInfoEndpointBinding(ProtocolEndpoint.BindingEnum userInfoEndpointBinding);
+
+        OIDCIdentityProviderBuilder setUserInfoEndpointUrl(String userInfoEndpointUrl);
+
+        OIDCIdentityProviderBuilder setJwksEndpointBinding(ProtocolEndpoint.BindingEnum jwksEndpointBinding);
+
+        OIDCIdentityProviderBuilder setJwksEndpointUrl(String jwksEndpointUrl);
+
+        OIDCIdentityProviderBuilder setScopes(List<String> scopes);
+
+        OIDCIdentityProviderBuilder setClientId(String clientId);
+
+        OIDCIdentityProviderBuilder setClientSecret(String clientSecret);
+
+        OIDCIdentityProviderBuilder setBaseUrl(String baseUrl);
+
+        OIDCIdentityProviderBuilder setMaxClockSkew(Integer maxClockSkew);
+
+        OIDCIdentityProviderBuilder setSubjectTemplate(String subjectTemplate);
+
+        OIDCIdentityProviderBuilder setPolicySubjectMatchType(PolicySubjectMatchType policySubjectMatchType);
+
+        IdentityProvider buildAndCreate(Client client);
+    }
+
+    interface GoogleIdentityProviderBuilder {
```

Comment:
It doesn't look like these interfaces are needed as they contain all the fields of `OIDCIdentityProviderBuilder` ?

Maybe just do something like:
```java
static OIDCIdentityProviderBuilder google() {
        return Classes.newInstance("com.okta.sdk.impl.resource.identity.provider.DefaultGoogleIdentityProviderBuilder");
    }

    static OIDCIdentityProviderBuilder facebook() {
        return Classes.newInstance("com.okta.sdk.impl.resource.identity.provider.DefaultFacebookIdentityProviderBuilder");
    }

    static OIDCIdentityProviderBuilder microsoft() {
        return Classes.newInstance("com.okta.sdk.impl.resource.identity.provider.DefaultMicrosoftIdentityProviderBuilder");
    }

    static OIDCIdentityProviderBuilder linkedin() {
        return Classes.newInstance("com.okta.sdk.impl.resource.identity.provider.DefaultLinkedInIdentityProviderBuilder");
    }
```

This would keep the actual implementation private (and not add an extra interface)

Reply:
Actually, `GoogleIdentityProviderBuilder` does not have to contain every method defined in `OIDCIdentityProviderBuilder`. Currently there are few methods that are shared between these interfaces. I wanted to keep these interfaces separate as it otherwise becomes an overhead to implement unnecessary methods from `OIDCIdentityProviderBuilder`. This does not fit the inheritance design principle and therefore this duplication. From caller's perspective, a `Google` specific instance could be built easily with:
  `IdentityProvider createdIdp = IdentityProviderBuilder.google()
            .setName("Mock Google IdP")
            .setScopes(["openid", "profile", "email"])
            .setClientId("your-client-id")
            .setClientSecret("your-client-secret")
            .setIsProfileMaster(true)
            .setMaxClockSkew(120000)
            .setUserNameTemplate("idpuser.email")
            .setPolicySubjectMatchType(PolicySubjectMatchType.USERNAME)
            .buildAndCreate(client)`

`GoogleIdentityProviderBuilder` interface and its implementation detail is actually hidden from the caller. The caller still has to rely on `IdentityProviderBuilder.google()` to get a Google specific builder.

Reply:
Ahh, i see.  You could end up with a base `IdentityProviderBuilder` and extend them for OIDC, Google, etc

Reply:
done.